File: bingo.rb
#!/usr/bin/env ruby
# bingo.rb
# encoding: UTF-8

require 'json'
require 'fileutils'
require 'set'

COLORS = {
  reset: "\e[0m",
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

class BingoCard
  attr_reader :numbers, :marked

  def initialize
    @numbers = Array.new(5) { Array.new(5, 0) }
    @marked = Array.new(5) { Array.new(5, false) }
    generate
  end

  def generate
    ranges = [[1,15],[16,30],[31,45],[46,60],[61,75]]
    ranges.each_with_index do |(low, high), col|
      nums = (low..high).to_a.shuffle
      5.times { |row| @numbers[row][col] = nums[row] }
    end
    @marked[2][2] = true
  end

  def mark(num)
    5.times do |r|
      5.times do |c|
        if @numbers[r][c] == num
          @marked[r][c] = true
          return true
        end
      end
    end
    false
  end

  def check_win
    # строки
    5.times do |r|
      return true if @marked[r].all?
    end
    # столбцы
    5.times do |c|
      ok = true
      5.times { |r| ok = false unless @marked[r][c] }
      return true if ok
    end
    # диагонали
    ok1 = ok2 = true
    5.times do |i|
      ok1 = false unless @marked[i][i]
      ok2 = false unless @marked[i][4-i]
    end
    ok1 || ok2
  end

  def display(hide=false)
    if hide
      5.times do |r|
        row = []
        5.times do |c|
          row << (@marked[r][c] ? colorize('XX', :green) : '??')
        end
        puts row.join(' ')
      end
      return
    end
    5.times do |r|
      row = []
      5.times do |c|
        if @marked[r][c]
          row << colorize(@numbers[r][c].to_s.rjust(2), :green)
        else
          row << @numbers[r][c].to_s.rjust(2)
        end
      end
      puts row.join(' ')
    end
  end
end

class BingoGame
  attr_reader :mode, :players, :called, :current

  def initialize(mode)
    @mode = mode
    @called = []
    @current = nil
    setup_players
  end

  def setup_players
    if @mode == 'single'
      @players = [
        { name: 'Игрок', card: BingoCard.new, won: false },
        { name: 'Компьютер', card: BingoCard.new, won: false }
      ]
    else
      @players = [
        { name: 'Игрок 1', card: BingoCard.new, won: false },
        { name: 'Игрок 2', card: BingoCard.new, won: false }
      ]
    end
  end

  def call_number
    available = (1..75).reject { |n| @called.include?(n) }
    return nil if available.empty?
    num = available.sample
    @called << num
    @current = num
    num
  end

  def mark_all(num)
    @players.each { |p| p[:card].mark(num) }
  end

  def check_winner
    @players.each_with_index do |p, i|
      return i if p[:card].check_win
    end
    nil
  end

  def display_state
    puts colorize('=' * 50, :bold)
    puts colorize("Вызванное число: #{@current || '—'}", :yellow)
    @players.each_with_index do |p, i|
      puts colorize("\n#{p[:name]}:", :bold)
      hide = (@mode == 'single' && i == 1)
      p[:card].display(hide)
    end
    puts colorize('=' * 50, :bold)
  end

  def play
    puts colorize('🎲 Добро пожаловать в Бинго!', :bold)
    if @mode == 'single'
      puts 'Игра против компьютера.'
    else
      puts 'Игра для двух игроков.'
    end
    puts 'Нажимайте Enter для вызова следующего числа.'
    puts 'Для выхода введите q.'

    loop do
      display_state
      print '> '
      cmd = gets.chomp.strip
      break if cmd == 'q'
      next unless cmd.empty?
      num = call_number
      if num.nil?
        puts colorize('Все числа вызваны! Ничья.', :yellow)
        break
      end
      mark_all(num)
      winner = check_winner
      if winner
        name = @players[winner][:name]
        puts colorize("🎉 Победитель: #{name}!", :green)
        display_state
        break
      end
      sleep 0.5
    end
  end
end

def main
  mode = 'single'
  show_stats = false
  reset_stats = false
  ARGV.each do |arg|
    case arg
    when 'two' then mode = 'two'
    when '-s', '--stats' then show_stats = true
    when '-r', '--reset' then reset_stats = true
    when '-h', '--help'
      puts 'Usage: ruby bingo.rb [single|two] [-s] [-r]'
      return
    end
  end
  if reset_stats
    f = File.join(Dir.home, '.bingo_stats.json')
    File.delete(f) if File.exist?(f)
    puts 'Статистика сброшена.'
    return
  end
  if show_stats
    puts 'Статистика не реализована в Ruby для простоты.'
    return
  end
  game = BingoGame.new(mode)
  game.play
end

main if __FILE__ == $0

File: bingo.py
# bingo.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import random
import json
import time
from pathlib import Path

# ANSI-цвета
COLORS = {
    'reset': '\033[0m',
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'bold': '\033[1m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

class BingoCard:
    def __init__(self):
        self.numbers = [[0]*5 for _ in range(5)]
        self.marked = [[False]*5 for _ in range(5)]
        self.generate()

    def generate(self):
        # Колонки: B:1-15, I:16-30, N:31-45, G:46-60, O:61-75
        ranges = [(1,15), (16,30), (31,45), (46,60), (61,75)]
        for col in range(5):
            nums = random.sample(range(ranges[col][0], ranges[col][1]+1), 5)
            for row in range(5):
                self.numbers[row][col] = nums[row]
        # Центр - FREE (всегда отмечен)
        self.marked[2][2] = True

    def mark(self, num):
        for r in range(5):
            for c in range(5):
                if self.numbers[r][c] == num:
                    self.marked[r][c] = True
                    return True
        return False

    def check_win(self):
        # Строки
        for r in range(5):
            if all(self.marked[r][c] for c in range(5)):
                return True, 'row', r
        # Столбцы
        for c in range(5):
            if all(self.marked[r][c] for r in range(5)):
                return True, 'col', c
        # Диагонали
        if all(self.marked[i][i] for i in range(5)):
            return True, 'diag', 0
        if all(self.marked[i][4-i] for i in range(5)):
            return True, 'diag', 1
        return False, None, None

    def display(self, hide=False):
        if hide:
            # Показываем только первый ряд (как в реальном бинго)
            # Но для простоты покажем всю карточку, но скроем цифры
            for r in range(5):
                row = []
                for c in range(5):
                    if self.marked[r][c]:
                        row.append(colorize('XX', 'green'))
                    else:
                        row.append('??')
                print(' '.join(row))
            return
        for r in range(5):
            row = []
            for c in range(5):
                if self.marked[r][c]:
                    row.append(colorize(f'{self.numbers[r][c]:2}', 'green'))
                else:
                    row.append(f'{self.numbers[r][c]:2}')
            print(' '.join(row))

class BingoGame:
    def __init__(self, mode='single'):
        self.mode = mode
        self.players = []
        self.called = []
        self.current_number = None
        self.stats_file = Path.home() / '.bingo_stats.json'
        self.stats = self.load_stats()
        self.setup_players()

    def load_stats(self):
        if self.stats_file.exists():
            with open(self.stats_file, 'r') as f:
                return json.load(f)
        return {}

    def save_stats(self):
        with open(self.stats_file, 'w') as f:
            json.dump(self.stats, f, indent=2)

    def setup_players(self):
        if self.mode == 'single':
            self.players = [('Игрок', BingoCard(), False), ('Компьютер', BingoCard(), False)]
        else:  # two
            self.players = [('Игрок 1', BingoCard(), False), ('Игрок 2', BingoCard(), False)]

    def call_number(self):
        available = [n for n in range(1, 76) if n not in self.called]
        if not available:
            return None
        num = random.choice(available)
        self.called.append(num)
        self.current_number = num
        return num

    def mark_all(self, num):
        for player in self.players:
            player[1].mark(num)

    def check_winner(self):
        for i, (name, card, won) in enumerate(self.players):
            if card.check_win()[0]:
                return i
        return None

    def display_state(self):
        print(colorize('='*50, 'bold'))
        print(colorize(f'Вызванное число: {self.current_number if self.current_number else "—"}', 'yellow'))
        for idx, (name, card, won) in enumerate(self.players):
            print(colorize(f'\n{name}:', 'bold'))
            if self.mode == 'single' and idx == 1:  # компьютер скрыт
                card.display(hide=True)
            else:
                card.display()
        print(colorize('='*50, 'bold'))

    def play(self):
        print(colorize('🎲 Добро пожаловать в Бинго!', 'bold'))
        if self.mode == 'single':
            print('Игра против компьютера.')
        else:
            print('Игра для двух игроков.')
        print('Нажимайте Enter для вызова следующего числа.')
        print('Для выхода введите q.\n')

        while True:
            self.display_state()
            cmd = input('> ').strip().lower()
            if cmd == 'q':
                print('Выход.')
                break
            if cmd != '':
                continue
            num = self.call_number()
            if num is None:
                print(colorize('Все числа вызваны! Ничья.', 'yellow'))
                break
            self.mark_all(num)
            winner = self.check_winner()
            if winner is not None:
                name = self.players[winner][0]
                print(colorize(f'🎉 Победитель: {name}!', 'green'))
                # Подсветка выигрышной линии
                self.display_state()
                # Обновление статистики
                if name not in self.stats:
                    self.stats[name] = {'wins': 0, 'losses': 0}
                self.stats[name]['wins'] += 1
                for i, (n, _, _) in enumerate(self.players):
                    if i != winner:
                        if n not in self.stats:
                            self.stats[n] = {'wins': 0, 'losses': 0}
                        self.stats[n]['losses'] += 1
                self.save_stats()
                print(colorize(f'Статистика сохранена.', 'blue'))
                break
            time.sleep(0.5)

def main():
    mode = 'single'
    show_stats = False
    reset_stats = False
    args = sys.argv[1:]
    for arg in args:
        if arg == 'two':
            mode = 'two'
        elif arg == '-s' or arg == '--stats':
            show_stats = True
        elif arg == '-r' or arg == '--reset':
            reset_stats = True
        elif arg == '-h' or arg == '--help':
            print('Usage: bingo.py [single|two] [-s] [-r]')
            return
    if reset_stats:
        stats_file = Path.home() / '.bingo_stats.json'
        if stats_file.exists():
            stats_file.unlink()
        print('Статистика сброшена.')
        return
    if show_stats:
        stats_file = Path.home() / '.bingo_stats.json'
        if stats_file.exists():
            with open(stats_file, 'r') as f:
                stats = json.load(f)
                print(colorize('📊 Статистика:', 'bold'))
                for name, data in stats.items():
                    print(f'  {name}: победы {data["wins"]}, поражения {data["losses"]}')
        else:
            print('Статистика пуста.')
        return
    game = BingoGame(mode)
    game.play()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print(colorize('\nИгра прервана.', 'yellow'))
        sys.exit(0)
